File: interface/controller/py/controller.py
# get tcp_protocol description (vscode[debuger] launched at root directory)
import sys
sys.path.append(r"./py")
import tcp_protocol_client as tcc

# init tcp host and port
HOST = 'localhost'
PORT = 9999

# get model description
import model
import model_enc

# get other tools
import numpy as np
import time

def origianl():
    # set simulation(this section have to set same with plant)
    sampling_time = 0.02
    run_signal = True

    # get model from model description file
    ctrl = model.ctrl(sampling_time)

    # input/output initialization
    y = np.array([[0],
                  [0]], dtype=float)
    u = np.array([[0]], dtype=float)

    with tcc.tcp_client(HOST, PORT) as tccp:
        while run_signal:
            # running signal send for controller
            _, signal = tccp.recv()

            if signal == "run":
                # get plant output
                _, y0 = tccp.recv()
                _, y1 = tccp.recv()
                y[0, 0] = y0
                y[1, 0] = y1

                # send control input data
                tccp.send(u[0, 0])

                # plant state update and generate output
                ctrl.state_update(y)
                u = ctrl.get_output()
            elif signal == "end":
                # end of loop signal get
                run_signal = False
                break

def arx():
    # set simulation(this section have to set same with plant)
    sampling_time = 0.02
    run_signal = True

    # get model from model description file
    ctrl = model.ctrl(sampling_time)
    ctrl_arx = model.ctrl_arx(ctrl.F, ctrl.G, ctrl.H, ctrl.J)

    # input/output initialization
    y = np.array([[0],
                  [0]], dtype=float)
    u = np.array([[0]], dtype=float)

    with tcc.tcp_client(HOST, PORT) as tccp:
        while run_signal:
            # running signal send for controller
            _, signal = tccp.recv()

            if signal == "run":
                # get plant output
                _, y0 = tccp.recv()
                _, y1 = tccp.recv()
                y[0, 0] = y0
                y[1, 0] = y1

                # send control input data
                tccp.send(u[0, 0])

                # arx memory update and generate output
                ctrl_arx.mem_update(y, u)
                u = ctrl_arx.get_output()
            elif signal == "end":
                # end of loop signal get
                run_signal = False
                break

def arx_q():
    # set simulation(this section have to set same with plant)
    sampling_time = 0.02
    run_signal = True

    # get model from model description file
    ctrl = model.ctrl(sampling_time)
    ctrl_arx = model.ctrl_arx(ctrl.F, ctrl.G, ctrl.H, ctrl.J)
    ctrl_arx_q = model.ctrl_arx_q(ctrl_arx.HG, ctrl_arx.HL)

    # set quantized level and quantize matrix
    ctrl_arx_q.set_level(1000, 1000)
    ctrl_arx_q.quantize()

    # input/output initialization
    y = np.array([[0],
                  [0]], dtype=float)
    u = np.array([[0]], dtype=float)

    with tcc.tcp_client(HOST, PORT) as tccp:
        while run_signal:
            # running signal send for controller
            _, signal = tccp.recv()

            if signal == "run":
                # get plant output
                _, y0 = tccp.recv()
                _, y1 = tccp.recv()
                y[0, 0] = y0
                y[1, 0] = y1

                # send control input data
                tccp.send(u[0, 0])

                # arx memory update and generate output
                ctrl_arx_q.mem_update(y, u)
                u = ctrl_arx_q.get_output()
            elif signal == "end":
                # end of loop signal get
                run_signal = False
                break

def arx_enc():
    # set simulation(this section have to set same with plant)
    sampling_time = 0.02
    run_signal = True

    # get model from model description file
    ctrl = model.ctrl(sampling_time)
    ctrl_arx = model.ctrl_arx(ctrl.F, ctrl.G, ctrl.H, ctrl.J)
    ctrl_arx_q = model.ctrl_arx_q(ctrl_arx.HG, ctrl_arx.HL)

    # set quantized level and quantize matrix
    ctrl_arx_q.set_level(1000, 1000)
    ctrl_arx_q.quantize()

    # get crypto model from model_enc
    crypto_cl = model_enc.crypto()
    enc_4_sys = model_enc.enc_for_system(crypto_cl, ctrl_arx_q.HG_q, ctrl_arx_q.HL_q)
    enc_4_sys.set_level(ctrl_arx_q.r, ctrl_arx_q.s)
    ctrl_enc = model_enc.ctrl_enc(crypto_cl.crypto_context, enc_4_sys.PQ_enc, enc_4_sys.Z_enc)

    # input/output initialization
    y = np.array([[0],
                  [0]], dtype=float)
    u = np.array([[0]], dtype=float)

    with tcc.tcp_client(HOST, PORT) as tccp:
        while run_signal:
            # running signal send for controller
            _, signal = tccp.recv()

            if signal == "run":
                # start clock set
                stc = time.perf_counter_ns()

                # get plant output
                _, y0 = tccp.recv()
                _, y1 = tccp.recv()
                y[0, 0] = y0
                y[1, 0] = y1

                # send control input data
                tccp.send(u[0, 0])

                # y and u value encryption after packing
                signal = enc_4_sys.enc_signal(y, u)
                
                ## controller description ##
                # ------------------------------------------------ #
                # ctrl mem update on encrypted space after encryption input/output value
                ctrl_enc.mem_update(signal)
                
                # get control input on ciphertext space
                enc_u = ctrl_enc.get_output()
                # ------------------------------------------------ #

                int_u = enc_4_sys.dec_signal(enc_u)

                u[0, 0] = float(int_u) / ctrl_arx_q.r / ctrl_arx_q.s

                # end clock set
                edc = time.perf_counter_ns()
                duration = (edc - stc) / 1000000
                print(f"loop time: {duration}ms")

                # ------------------------------------------------ #
            elif signal == "end":
                # end of loop signal get
                run_signal = False
                break

def main():
    # origianl()
    # arx()
    # arx_q()
    arx_enc()
    
if __name__ == "__main__":
    main()

File: interface/controller/cpp/controller.cpp
// get tcp_protocol decription
#include "../../../cpp/tcp_protocol_client.h"

// get other tools
#include <iostream>
#include <string>
#include <chrono>
using namespace std;

// get model description
#include "model_enc.h"

// init tcp host and port
const string host = "127.0.0.1";
const int port = 9999;

int main()
{
    // set simulation(this section have to set same with plant)
    double samplint_time = 0.03;
    bool run_signal = true;

    // get crypto model from model_enc.h
    crypto crypto_cl = crypto();
    enc_for_system enc_4_sys = enc_for_system(crypto_cl);
    enc_4_sys.set_level(1000, 1000);
    ctrl_enc ctrl_enc_v = ctrl_enc(crypto_cl.get_crypto(), crypto_cl.get_relinkey(), crypto_cl.get_galoiskeys());
    ctrl_enc_v.set_pq(enc_4_sys.get_PQ_enc());
    ctrl_enc_v.set_io(enc_4_sys.get_Z_enc());

    // set tcp client
    tcp_client tccp = tcp_client(host, port);
    string signal;

    // for check cycle time
    auto stc = chrono::high_resolution_clock::now();
    auto edc = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
    double run_time;

    // input/output initialization
    vector<double> y(2, 0.0);
    vector<double> u(1, 0.0);
    vector<int64_t> int_u(1, 0LL);

    while(run_signal)
    {
        signal = tccp.Recv<string>();

        if(signal == "run")
        {
            // start clock set
            stc = chrono::high_resolution_clock::now();

            // get plant output
            double y0 = tccp.Recv<double>();
            double y1 = tccp.Recv<double>();
            y[0] = y0;
            y[1] = y1;
            
            // send control input data
            tccp.Send<double>(u[0]);

            // y and u value encryption after packing
            Ciphertext signal = enc_4_sys.enc_signal(y, u);

            // -- controller description -- //
            // ========================================================== //
            // ctrl mem update on encrypted space after encryption input/output value
            ctrl_enc_v.mem_update(signal);

            // get control input on ciphertext space
            Ciphertext enc_u = ctrl_enc_v.get_output();
            // ========================================================== //

            int_u = enc_4_sys.dec_signal(enc_u); 

            u[0] = (double)(int_u[0]) / enc_4_sys.get_level()[0] / enc_4_sys.get_level()[1];

            // end clock set
            edc = chrono::high_resolution_clock::now();
            duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
            run_time = duration.count() / 1000000;
            cout << "loop time: " << run_time << "ms" << endl;

        }
        else if(signal == "end")
        {
            // end of loop signal get
            run_signal = false;
            break;
        }
    }

    return 0;
}
